### PR TITLE
SINGA-474 hardsigmoid operator

### PR DESCRIPTION
**Pull request by ShichengChen**
Implement ONNX Operators following https://github.com/onnx/onnx/blob/master/docs/Operators.md

### CODE REVIEW

File: python/singa/autograd.py [hunk 1]
```diff
@@ -358,6 +358,51 @@ def relu(x):
...
+        """
+        if training:
+            self.input = x
+
+        x = singa.AddFloat(singa.MultFloat(x,self.alpha),self.gamma)
+        x = singa.ReLU(x)
+        mask = singa.LTFloat(x, 1.0)
+        mask2 = singa.GEFloat(x, 1.0)
+
+        ans = singa.__add__(singa.__mul__(x, mask),mask2)
```
Comment (nudles) [inline python/singa/autograd.py]: why add the mask2 here?

File: test/python/test_operation.py [hunk 2]
```diff
@@ -321,7 +321,34 @@ def test_LeakyRelu(self):
...
+        a=0.5
+        g=0.6
+        y = np.clip(x * 0.5 + 0.6, 0, 1)
+        x = tensor.from_numpy(x)
+        x.to_device(cpu_dev)
+
+        result = autograd.hardsigmoid(x,a,g)
+        dx = result.creator.backward(x.data)
+
+        np.testing.assert_array_almost_equal(tensor.to_numpy(result), y, decimal=5)
```
Comment (nudles) [inline test/python/test_operation.py]: backward is not tested..

File: python/singa/autograd.py (hunk 1 above)
Comment (ShichengChen) [reply python/singa/autograd.py]: if x >=1 then x == 1
I first made x[x>=1]=0 and then x[x>=1]+=1

File: test/python/test_operation.py (hunk 2 above)
Comment (ShichengChen) [reply test/python/test_operation.py]: yes, I just added test

File: python/singa/autograd.py (hunk 1 above)
Comment (joddiy) [reply python/singa/autograd.py]: can we write like this?
```
x = singa.AddFloat(singa.MultFloat(x, self.alpha), self.gamma)
x = singa.__clamp__(x, 0.0, 1.0)
```

File: python/singa/autograd.py [hunk 3]
```diff
@@ -358,6 +358,51 @@ def relu(x):
...
+
+    def backward(self, dy):
+        """
+        Args:
+            dy (CTensor): data for the dL / dy, L is the loss
+        Returns:
+            a (CTensor) dx
+        """
+
+        x = singa.AddFloat(singa.MultFloat(self.input,self.alpha),self.gamma)
```
[the remark this answers is not in the document]
Comment (ShichengChen) [reply python/singa/autograd.py]: yes, I just cached. We do not have clip operation now

File: python/singa/autograd.py (hunk 1 above)
Comment (nudles) [reply python/singa/autograd.py]: relu will turn x into x if x>=0 and 0 otherwise;
then, you do not need to add 1?

**Comment by nudles** [comment]
pls resolve the conflicts

**Pull request closed without merging**